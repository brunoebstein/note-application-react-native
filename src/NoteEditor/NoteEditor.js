import React, { Component } from 'react';
import { View, TextInput, StyleSheet, Button } from 'react-native';
import PropTypes from 'prop-types';
import isEqual from 'lodash.isequal';

export class NoteEditor extends Component {
  static propTypes = {
    note: PropTypes.shape({
      title: PropTypes.string.isRequired,
      content: PropTypes.string.isRequired,
    }),
    saveNote: PropTypes.func,
    onSavedNote: PropTypes.func,
    onCancel: PropTypes.func,
  };

  constructor(props) {
    super(props);

    const note = props.note;
    this.state = { note };
  }

  componentWillReceiveProps(newProps) {
    const note = newProps.note;
    this.setState({ note });
  }

  render() {
    const { note: propsNote, onCancel } = this.props;
    const { note: stateNote } = this.state;
    const { title, content } = stateNote || {
      title: undefined,
      content: undefined,
    };

    return (
      <View style={styles.container}>
        <TextInput
          placeholder="Title"
          defaultValue={title}
          onChangeText={this._updateTitle}
        />
        <View style={styles.contentInput}>
          <TextInput
            placeholder="Content"
            multiline={true}
            defaultValue={content}
            onChangeText={this._updateContent}
          />
        </View>
        <View style={styles.buttonGroup}>
          <Button
            onPress={() => {
              if (typeof onCancel === 'function') {
                onCancel();
              }
            }}
            title="Cancel"
          />
          <Button
            onPress={this.save}
            title="Save"
            disabled={isEqual(stateNote, propsNote)}
          />
        </View>
      </View>
    );
  }

  _updateTitle = title => {
    this.setState(state => ({
      ...state,
      note: {
        ...state.note,
        title,
      },
    }));
  };

  _updateContent = content => {
    this.setState(state => ({
      ...state,
      note: {
        ...state.note,
        content,
      },
    }));
  };

  save = () => {
    if (typeof this.props.saveNote !== 'function') return;

    this.props.saveNote(this.state.note);

    if (typeof this.props.onSavedNote === 'function') {
      this.props.onSavedNote();
    }

    this.setState(state => ({
      ...state,
      note: {},
    }));
  };
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: 20,
  },
  contentInput: {
    flex: 1,
    justifyContent: 'flex-start',
  },
  buttonGroup: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    padding: 20,
  },
});
