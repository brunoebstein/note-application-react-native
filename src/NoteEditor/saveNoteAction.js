import { SAVE_NOTE } from '../actionTypes';

export const saveNote = note => ({
  type: SAVE_NOTE,
  payload: note,
});
