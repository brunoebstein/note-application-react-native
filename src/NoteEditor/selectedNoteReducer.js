const defaultState = null;

export const selectedNoteReducer = (state = defaultState, action) => {
  switch (action && action.type) {
    case 'SAVE_NOTE':
      return null;
    default:
      return state;
  }
};
