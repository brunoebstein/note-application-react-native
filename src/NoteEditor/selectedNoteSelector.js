export const selectedNoteSelector = state => state.selectedNote;
