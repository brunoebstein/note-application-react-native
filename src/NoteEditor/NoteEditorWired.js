import { connect } from 'react-redux';

import { selectedNoteSelector } from './selectedNoteSelector';
import { saveNote } from './saveNoteAction';
import { NoteEditor } from './NoteEditor';

const mapStateToProps = state => ({
  note: selectedNoteSelector(state),
});

export const NoteEditorWired = connect(
  mapStateToProps,
  { saveNote },
)(NoteEditor);
