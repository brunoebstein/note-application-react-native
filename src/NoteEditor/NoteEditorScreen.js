import React, { Component } from 'react';

import { themedNavigationOptions } from '../NoteNavigator/themedNavigationOptions';
import { NoteEditorWired } from '../NoteEditor/NoteEditorWired';

export class NoteEditorScreen extends Component {
  static navigationOptions = themedNavigationOptions({
    title: 'Edit note',
  });

  render() {
    const {
      navigation: { navigate, goBack },
    } = this.props;
    return (
      <NoteEditorWired
        onSavedNote={() => navigate('NoteListView')}
        onCancel={() => goBack()}
      />
    );
  }
}
