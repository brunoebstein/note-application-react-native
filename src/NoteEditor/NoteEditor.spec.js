import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { NoteEditor } from './NoteEditor';

describe('a NoteEditor', () => {
  const note = {
    title: 'test title',
    content: 'test content',
  };

  describe('simple render', () => {
    it('should display the editor form when no Note', () => {
      const editor = renderer.create(<NoteEditor />).toJSON();
      expect(editor).toMatchSnapshot();
    });

    it('should display the editor form for an existing note', () => {
      const editor = renderer.create(<NoteEditor note={note} />).toJSON();
      expect(editor).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should have a disabled save button', () => {
      const editor = shallow(<NoteEditor note={note} />);

      expect(
        editor
          .find('Button')
          .last()
          .prop('disabled'),
      ).toEqual(true);
    });

    it('should enable the save button when title change', () => {
      const editor = shallow(<NoteEditor note={note} />);

      editor.setState({
        note: {
          title: 'test title changed',
          content: 'test content',
        },
      });

      expect(
        editor
          .find('Button')
          .last()
          .prop('disabled'),
      ).toEqual(false);
    });

    it('should enable the save button when content change', () => {
      const editor = shallow(<NoteEditor note={note} />);

      editor.setState({
        note: {
          title: 'test title',
          content: 'test content changed',
        },
      });

      expect(
        editor
          .find('Button')
          .last()
          .prop('disabled'),
      ).toEqual(false);
    });
  });
});
