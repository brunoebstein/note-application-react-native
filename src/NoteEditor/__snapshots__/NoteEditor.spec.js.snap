// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a NoteEditor simple render should display the editor form for an existing note 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginTop": 20,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    defaultValue="test title"
    onChangeText={[Function]}
    placeholder="Title"
  />
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      defaultValue="test content"
      multiline={true}
      placeholder="Content"
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Cancel
        </Text>
      </View>
    </View>
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
          "disabled",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#cdcdcd",
              },
            ]
          }
        >
          Save
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`a NoteEditor simple render should display the editor form when no Note 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginTop": 20,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Title"
  />
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      multiline={true}
      placeholder="Content"
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Cancel
        </Text>
      </View>
    </View>
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
          "disabled",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#cdcdcd",
              },
            ]
          }
        >
          Save
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`a NoteEditor simple render should display the editor form for an existing note 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginTop": 20,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    defaultValue="test title"
    onChangeText={[Function]}
    placeholder="Title"
  />
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      defaultValue="test content"
      multiline={true}
      onChangeText={[Function]}
      placeholder="Content"
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Cancel
        </Text>
      </View>
    </View>
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
          "disabled",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#cdcdcd",
              },
            ]
          }
        >
          Save
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`a NoteEditor simple render should display the editor form when no Note 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginTop": 20,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Title"
  />
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      multiline={true}
      onChangeText={[Function]}
      placeholder="Content"
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Cancel
        </Text>
      </View>
    </View>
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
          "disabled",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#cdcdcd",
              },
            ]
          }
        >
          Save
        </Text>
      </View>
    </View>
  </View>
</View>
`;
